File: src/main/java/com/github/sorusclient/client/bootstrap/loader/ClassLoader.java
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.bootstrap.loader;

import lombok.SneakyThrows;
import lombok.val;
import org.apache.commons.io.IOUtils;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Enumeration;
import java.util.List;

public class ClassLoader extends URLClassLoader {

    private final List<Object> transformers = new ArrayList<>();
    private final Method canTransformMethod;
    private final Method transformMethod;
    private final List<String> exclusions = new ArrayList<>();
    private final List<URL> urls = new ArrayList<>();
    private final java.lang.ClassLoader realParent = ClassLoader.class.getClassLoader();

    @SneakyThrows
    public ClassLoader() {
        super(new URL[0], null);

        exclusions.add("java.");
        exclusions.add("jdk.");
        exclusions.add("javax.");
        exclusions.add("sun.");
        exclusions.add("com.sun.");
        exclusions.add("org.xml.");
        exclusions.add("org.w3c.");
        exclusions.add("org.apache.");
        exclusions.add("org.slf4j.");
        exclusions.add("com.mojang.blocklist.");
        exclusions.add("com.github.sorusclient.client.bootstrap.transformer.Transformer");

        canTransformMethod = this.loadClass("com.github.sorusclient.client.bootstrap.transformer.Transformer").getMethod("canTransform", String.class);
        transformMethod = this.loadClass("com.github.sorusclient.client.bootstrap.transformer.Transformer").getMethod("transform", String.class, byte[].class);
    }

    @Override
    public Class<?> loadClass(String name) throws ClassNotFoundException {
        for (val exclusion : exclusions) {
            if (name.startsWith(exclusion)) {
                try {
                    return realParent.loadClass(name);
                } catch (Exception ignored) {

                }
            }
        }

        var clazz = super.findLoadedClass(name);
        if (clazz == null) {
            val data = getModifiedBytes(name);
            clazz = super.defineClass(name, data, 0, data.length);
        }
        return clazz;
    }

    @Override
    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        val clazz = this.loadClass(name);
        if (resolve) {
            resolveClass(clazz);
        }
        return clazz;
    }

    @SneakyThrows({IllegalAccessException.class, InvocationTargetException.class})
    private byte[] getModifiedBytes(String name) throws ClassNotFoundException {
            var data = loadClassData(name);

            if (data.length == 0) {
                throw new ClassNotFoundException(name);
            }

            for (val transformer : getTransformers()) {
                val formattedName = name.replace(".", "/");

                if ((boolean) canTransformMethod.invoke(transformer, formattedName)) {
                    data = (byte[]) transformMethod.invoke(transformer, formattedName, data);
                }
            }

            if (data.length == 0) {
                throw new ClassNotFoundException(name);
            }
            return data;
    }

    private List<Object> getTransformers() {
        return transformers;
    }

    @SneakyThrows
    private byte[] loadClassData(String className) {
        val resources = getResources(className.replace(".", "/") + ".class");
        final List<byte[]> datas = new ArrayList<>();

        while (resources.hasMoreElements()) {
            val resource = resources.nextElement();
            datas.add(IOUtils.toByteArray(resource));
        }

        if (datas.size() > 0) {
            return datas.toArray(new byte[0][])[0];
        } else {
            return new byte[0];
        }
    }

    @SneakyThrows
    @Override
    public Enumeration<URL> getResources(String name) {
        final List<URL> parentResources  = Collections.list(realParent.getResources(name));
        final List<URL> filteredURLs = new ArrayList<>(parentResources);
        for (var pathUrl : Collections.list(findResources(name))) {
            for (var url : urls) {
                if (pathUrl.getFile().contains(url.getFile())) {
                    filteredURLs.add(pathUrl);
                }
            }
        }

        return Collections.enumeration(filteredURLs);
    }

    @SneakyThrows
    @Override
    public URL getResource(String name) {
        val resources = getResources(name);
        if (resources.hasMoreElements()) {
            return resources.nextElement();
        }
        return realParent.getResource(name);
    }

    @Override
    public void addURL(URL url) {
        urls.add(url);
        super.addURL(url);
    }

    public void removeURL(URL url) {
        urls.remove(url);
    }

    @SneakyThrows
    public void addTransformer(Class<?> transformer) {
        this.transformers.add(transformer.getConstructor().newInstance());
    }

}

File: src/main/kotlin/com/github/sorusclient/client/bootstrap/loader/ClassLoader.kt
package com.github.sorusclient.client.bootstrap.loader

import org.apache.commons.io.IOUtils
import java.io.IOException
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.net.URL
import java.net.URLClassLoader
import java.util.*
import kotlin.collections.ArrayList

class ClassLoader : URLClassLoader(arrayOfNulls(0), null) {

    private val transformers: MutableList<Any> = ArrayList()
    private val canTransformMethod: Method
    private val transformMethod: Method
    private val exclusions: MutableList<String> = ArrayList()
    private val urls: MutableList<URL> = ArrayList()
    private val realParent = ClassLoader::class.java.classLoader

    init {
        exclusions.add("java.")
        exclusions.add("jdk.")
        exclusions.add("javax.")
        exclusions.add("sun.")
        exclusions.add("com.sun.")
        exclusions.add("org.xml.")
        exclusions.add("org.w3c.")
        exclusions.add("org.apache.")
        exclusions.add("org.slf4j.")
        exclusions.add("com.mojang.blocklist.")
        exclusions.add("com.github.sorusclient.client.bootstrap.transformer.Transformer")
        exclusions.add("com.github.glassmc.loader.api.ClassDefinition")

        canTransformMethod = this.loadClass("com.github.sorusclient.client.bootstrap.transformer.Transformer")!!.getMethod(
            "canTransform",
            String::class.java
        )

        transformMethod = this.loadClass("com.github.sorusclient.client.bootstrap.transformer.Transformer")!!.getMethod(
            "transform",
            String::class.java,
            ByteArray::class.java
        )
    }

    @Throws(ClassNotFoundException::class)
    override fun loadClass(name: String): Class<*>? {
        for (exclusion in exclusions) {
            if (name.startsWith(exclusion)) {
                try {
                    return realParent.loadClass(name)
                } catch (ignored: ClassNotFoundException) {
                }
            }
        }
        var clazz = super.findLoadedClass(name)
        if (clazz == null) {
            val data = getModifiedBytes(name)
            clazz = super.defineClass(name, data, 0, data.size)
        }
        return clazz
    }

    @Throws(ClassNotFoundException::class)
    override fun loadClass(name: String, resolve: Boolean): Class<*>? {
        val clazz = this.loadClass(name)
        if (resolve) {
            resolveClass(clazz)
        }
        return clazz
    }

    @Throws(ClassNotFoundException::class)
    fun getModifiedBytes(name: String): ByteArray {
        var data = loadClassData(name)
        for (transformer in getTransformers()) {
            val formattedName = name.replace(".", "/")

            if (canTransformMethod.invoke(transformer, formattedName) as Boolean) {
                data = transformMethod.invoke(transformer, formattedName, data) as ByteArray
            }
        }

        if (data.isEmpty()) {
            throw ClassNotFoundException(name)
        }
        return data
    }

    private fun getTransformers(): List<Any> {
        //val transformers: MutableList<Any> = ArrayList()
        //transformers.addAll(this.transformers[0])
        //transformers.addAll(this.transformers[1])
        //transformers.addAll(this.transformers[2])
        return this.transformers
    }

    private fun loadClassData(className: String): ByteArray {
        return try {
            val resources = getResources(className.replace(".", "/") + ".class")
            val locations: MutableList<String> = ArrayList()
            val datas: MutableList<ByteArray> = ArrayList()
            while (resources.hasMoreElements()) {
                val resource = resources.nextElement()
                locations.add(resource.toString())
                datas.add(IOUtils.toByteArray(resource))
            }

            val result: Array<ByteArray> = if (datas.size > 0) {
                datas.toTypedArray()
            } else {
                arrayOf(ByteArray(0))
            }

            result[0]
        } catch (e: IOException) {
            e.printStackTrace()
            ByteArray(0)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
            ByteArray(0)
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
            ByteArray(0)
        }
    }

    @Throws(IOException::class)
    override fun getResources(name: String): Enumeration<URL> {
        val parentResources: List<URL> = Collections.list(realParent.getResources(name))
        val filteredURLs: MutableList<URL> = ArrayList(parentResources)
        for (pathUrl in Collections.list(findResources(name))) {
            for (url in urls) {
                if (pathUrl.file.contains(url.file)) {
                    filteredURLs.add(pathUrl)
                }
            }
        }
        return Collections.enumeration(filteredURLs)
    }

    override fun getResource(name: String): URL? {
        try {
            val resources = getResources(name)
            if (resources.hasMoreElements()) {
                return resources.nextElement()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return realParent.getResource(name)
    }

    public override fun addURL(url: URL) {
        urls.add(url)
        super.addURL(url)
    }

    fun removeURL(url: URL) {
        urls.remove(url)
    }

    fun addTransformer(transformer: Class<*>) {
        this.transformers.add(transformer.getConstructor().newInstance())
        /*try {
            val listIndex: Int
            listIndex = when (order) {
                TransformerOrder.FIRST -> 0
                TransformerOrder.LAST -> 2
                TransformerOrder.DEFAULT -> 1
                else -> -1
            }
            transformers[listIndex].add(transformer.getConstructor().newInstance())
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        }*/
    }
}